sematic/external_resource.py

```diff
@@ -0,0 +1,331 @@
+# Standard Library
+import logging
+import time
+import uuid
+from dataclasses import dataclass, field, fields, is_dataclass, replace
+from enum import Enum, unique
+from typing import TypeVar, final
+
+# Sematic
+from sematic.abstract_future import AbstractFuture
+from sematic.future_context import SematicContext, context
+from sematic.types.registry import register_type_assertion
+from sematic.utils.exceptions import (
+    IllegalStateTransitionError,
+    IllegalUseOfFutureError,
+    NotInSematicFuncError,
+)
+
+logger = logging.getLogger(__name__)
+
+
+@unique
+class ResourceState(Enum):
+    """Represents the state of an external resource.
+
+    Attributes
+    ----------
+    CREATED:
+        The object representing the resource has been instantiated in-memory
+        but no logic to allocate the resource has been executed.
+    ACTIVATING:
+        The resource is being allocated. This may take time. The resource cannot
+        be used yet.
+    ACTIVE:
+        The resource is allocated and ready for use.
+    DEACTIVATING:
+        The resource is being deactivated, and likely is no longer usable.
+    DEACTIVATED:
+        The resource has been deactivated and may not even exist anymore.
+        It is not usable.
+    """
+
+    CREATED = "CREATED"
+    ACTIVATING = "ACTIVATING"
+    ACTIVE = "ACTIVE"
+    DEACTIVATING = "DEACTIVATING"
+    DEACTIVATED = "DEACTIVATED"
+
+    def is_allowed_transition(self, other_state: "ResourceState") -> bool:
```

Comment:
Same comment as on the original PR. We now have three state machines (run/future, resolution, external resource). We may need an abstraction to DRY state transition validation logic.
Obviously allowed transitions will still need to be defined on each state machine.

Reply:
filed #408